docs/localization.md

```diff
@@ -0,0 +1,116 @@
+# Writing Localizeable Code
+
+As part of Microsoft's commitment to global-ready products, our developer tools are localized into 14 different languages.
+For this effort, vcpkg has hooks in order to allow our localization team to translate all messages that are shown to the user by the vcpkg tool.
+As a developer on vcpkg, the only thing you have to consider is the messages in the C++ source files,
+and `locales/messages.json` -- everything else will be generated and modified by the localization team.
+
+## Declaring a Message
+
+The process of writing a message to be displayed to the user starts with the message; if it needs to be in a header,
+use the `DECLARE_MESSAGE(<message-name>, <parameters>, <comment>, <english-message>` in that header,
+followed by `REGISTER_MESSAGE(<message-name>)` in a source file; if it can just be in a source file,
+use `DECLARE_AND_REGISTER_MESSAGE()`.
+
+### Message Names
+
+Message names should be descriptive and unique, and must be `CamelCase`.
+When referring to a message in `msg::format`, you must add `msg` to the front of it; for example,
+`DECLARE_MESSAGE(MyMessage, ...)` can be referred to as `msgMyMessage`.
+
+### Parameters
+
+Each placeholder in a localized format string must have a corresponding parameter in this list.
+It should look something like `(msg::arg1, msg::arg2)`.
+This allows us to make sure that we don't have format errors, where one forgets a placeholder,
+or adds an extra one.
+Each parameter in the list should be used in the message.
+
+### Comment
+
+The comment is used to give context to the translator as to what a placeholder will be replaced with.
+For example, in `BuildResultSummaryHeader` (`"SUMMARY FOR {triplet}"`),
+it will note that an example of `{triplet}` is `'x64-windows'`.
+Most message placeholders (see `messages.h` for the full list - search for `DECLARE_MSG_ARG`) have comments associated with them already.
+Only general placeholders like `{value}`, `{expected}`, and `{actual}` don't have comments associated with them.
+If you use any of these placeholders, write `example of {<name>} is '<expected-value>'` for each of those placeholders,
+separated by `\n`. You can also add context if you feel it's necessary as the first line.
```

Comment:
This part is a bit hard to follow. Maybe we should start with an example of an existing message that uses all parameters and break down each part of the declaration.